src/composed/message/types.rs

```diff
@@ -164,19 +164,88 @@ impl Edata {
         debug!("decrypting protected = {:?}", protected);
 
-        let decrypted_packet: &[u8] = if protected {
-            alg.decrypt_protected(&key, &mut res)
-        } else {
-            alg.decrypt(&key, &mut res)
-        }?;
+        match key {
+            PlainSessionKey::V4 { sym_alg, key } => {
+                ensure!(
+                    sym_alg != SymmetricKeyAlgorithm::Plaintext,
+                    "session key algorithm cannot be plaintext"
+                );
+
+                match self {
+                    Self::SymEncryptedProtectedData(p) => {
+                        ensure_eq!(
+                            self.version(),
+                            Some(1),
+                            "Version missmatch between key and integrity packet"
+                        );
+                        let data = p.decrypt(&key, Some(sym_alg))?;
+                        Message::from_bytes(Cursor::new(data))
+                    }
+                    Self::SymEncryptedData(p) => {
+                        ensure_eq!(
+                            self.version(),
+                            None,
+                            "Version missmatch between key and integrity packet"
+                        );
+                        let mut data = self.data()[..].to_vec();
+                        let res = sym_alg.decrypt(&key, &mut data)?;
+                        let l = res.len();
+                        drop(res);
+                        data.truncate(l);
+                        Message::from_bytes(Cursor::new(data))
+                    }
+                }
+            }
+            PlainSessionKey::V5 { key } => {
+                if protected {
+                    ensure_eq!(
+                        self.version(),
+                        Some(2),
+                        "Version missmatch between key and integrity packet"
+                    );
+                    unimplemented_err!("V5 decryption");
+                } else {
+                    // TODO: is this a valid case?
+                    ensure_eq!(
+                        self.version(),
+                        None,
+                        "Version missmatch between key and integrity packet"
+                    );
+                    unimplemented_err!("V5 decryption not protected");
+                }
+            }
+            PlainSessionKey::V6 { key } => {
+                match self {
+                    Self::SymEncryptedProtectedData(p) => {
+                        let decrypted_packet = p.decrypt(&key, None)?;
+
+                        // TODO: handle multiple messages (padding packets can appear here)
 
-        Message::from_bytes(Cursor::new(decrypted_packet.to_vec()))
+                        Message::from_bytes(Cursor::new(decrypted_packet))
+                    }
+                    Self::SymEncryptedData(_) => {
+                        // TODO: is this a valid case?
```

Comment:
I don't think this is a valid case:

> A v6 SKESK packet MUST NOT precede a v1 SEIPD packet or a deprecated Symmetrically Encrypted Data packet (see [Section 10.3.2.1](https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-13.html#encrypted-message-versions)).

(from https://www.ietf.org/archive/id/draft-ietf-openpgp-crypto-refresh-13.html#name-version-6-symmetric-key-enc)

So `SymEncryptedProtectedData` with version one must also be rejected here.